Add  annotation to allow using custom CDI bean methods as permission checkers

michalvavrik opened the pull request:
- Closes: https://github.com/quarkusio/quarkus/issues/43238

This PR is ready for review, however CI cannot run until https://github.com/quarkusio/quarkus-security/pull/56 is merged and Quarkus Security API is released and bumped in the Quarkus project.

sberyozkin commented:
Thanks @michalvavrik, sorry for the delay, will start looking tomorrow, need to sign off now

michalvavrik commented:
> Thanks @michalvavrik, sorry for the delay, will start looking tomorrow, need to sign off now

no worry, if don't have API anyway (I run all the related tests locally, so I don't expect CI issues), take your time, this won't be easy to review even though this PR consist mostly of tests

sberyozkin commented on docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc:
```diff
@@ -1055,8 +1055,83 @@ Because `MediaLibrary` is the `TvLibrary` class parent, a user with the `admin`
+
+@ApplicationScoped
+public class HelloPermissionChecker {
+
```
It is not used, I'd just remove RoutingContext.

sberyozkin commented on docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc:
```diff
@@ -1055,8 +1055,83 @@ Because `MediaLibrary` is the `TvLibrary` class parent, a user with the `admin`
+[[permission-checker]]
+==== Create permission checkers
+
+Previous examples required that the `SecurityIdentity` possessed required permissions.
```
```suggestion
By default, `SecurityIdentity` must be augmented with permissions which can be used  to check if this identity passes `@PermissionAllowed` authorisation restrictions.
```

sberyozkin commented on docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc:
```diff
@@ -1055,8 +1055,83 @@ Because `MediaLibrary` is the `TvLibrary` class parent, a user with the `admin`
+
+Previous examples required that the `SecurityIdentity` possessed required permissions.
+We added these permissions in the `SecurityIdentityAugmentor` and used role-to-permission mapping.
+Another way to check permissions required by the `@PermissionsAllowed` security annotation is a permission checker.
```
```suggestion
Alternatively, you can use a `@PermissionChecker` annotation to mark any CDI bean method as a permission checker.
```

sberyozkin commented on docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc:
```diff
@@ -1055,8 +1055,83 @@ Because `MediaLibrary` is the `TvLibrary` class parent, a user with the `admin`
+Previous examples required that the `SecurityIdentity` possessed required permissions.
+We added these permissions in the `SecurityIdentityAugmentor` and used role-to-permission mapping.
+Another way to check permissions required by the `@PermissionsAllowed` security annotation is a permission checker.
+The permission checkers are CDI bean methods annotated with a `@PermissionChecker`.
```
```suggestion
```

sberyozkin commented on docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc:
```diff
@@ -1055,8 +1055,83 @@ Because `MediaLibrary` is the `TvLibrary` class parent, a user with the `admin`
+We added these permissions in the `SecurityIdentityAugmentor` and used role-to-permission mapping.
+Another way to check permissions required by the `@PermissionsAllowed` security annotation is a permission checker.
+The permission checkers are CDI bean methods annotated with a `@PermissionChecker`.
+For example, a permission checker can be created like this:
```
I feel, given how significant this enhancement is, that we provide more intersting examples instead of say hello 🙂, IMHO, doing an `owner` example would be better.

michalvavrik replied:
I copied and tweak example in the https://github.com/quarkusio/quarkus/issues/43238 description. I did not include all these classes there as it seems like unnecessary noise. We need to show what the `@PermissionChecker` can do, capabilities, not ideas how to use it. Examples and docs is highly opinioned, let's suggest better example, I am completely fine with it. Thanks

sberyozkin commented on extensions/security/runtime/src/main/java/io/quarkus/security/runtime/QuarkusPermission.java:
```diff
@@ -0,0 +1,147 @@
+import io.smallrye.mutiny.Uni;
+
+/**
+ * Special type of the {@link PermissionsAllowed#permission()} that does not require {@link SecurityIdentity}
```
IMHO this is too implementation detail specific where you imply users can actually use it, and then, later, say users should not use it.
Can we just say `Special type of 'Permission' which is used by Quarkus Security to call CDI bean methods annotated with '{@link PermissionsChecker}' ` ? IMHO users don't need to know at the moment how this permission works internally

sberyozkin replied:
Michal, @michalvavrik, I hope you will be around for a long while :-), so you will explain if someone will want clarifications :-).

sberyozkin commented on extensions/security/deployment/src/test/java/io/quarkus/security/test/permissionsallowed/checker/NoArgsPermissionCheckerValidationFailureTest.java:
```diff
@@ -0,0 +1,48 @@
+import io.quarkus.security.PermissionsAllowed;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class NoArgsPermissionCheckerValidationFailureTest {
```
Very comprehensive test coverage, this is great.
Do you have a test which verifies that a checker which does not accept either `SecurityIdentity` or one of the JAX-RS method parameters is ignored ?

michalvavrik replied:
> Do you have a test which verifies that a checker which does not accept either SecurityIdentity or one of the JAX-RS method parameters is ignored ?

It's this one (though "not ignored" but leads to build failure). Maybe I am missing something? If you mean a different scenario, please give me a hint.

TBH this test you commented on only exists because you explicitly requested it, personally I would allow no args because you can also inject `SecurityIdentity`, so if someone has bean with 5 checkers and they all use `SecurityIdentity` and he prefers to use `@Inject` instead of method argument, I would not prevent it. But this is such a small detail I suppose it is not worth any discussion at all. Let's keep it.

sberyozkin replied:
Looks like it is this one, https://github.com/quarkusio/quarkus/pull/43846/files#diff-308cccee37a7dec2fc161afdbbfbbc2b922a91de72d6d1312faabf468467d550 ?

michalvavrik replied:
ah, ok, you meant something different; you wanted to validate that all the unknown params are reported, yes, it is always reported as when we cannot find match on the side of Permission constructor, build fails.

sberyozkin replied:
@michalvavrik See, we need some kind of evidence that the authenticated request can be authorized in that permission checker method. Otherwise users may end up doing some kind of self-authorization; there is no guarantee what that checker method will do it all correctly anyway but at least Quarkus will know the request specific info is available for that method.

sberyozkin replied:
@michalvavrik What is going to happen here:
```
@PermissionAllowed("id")
void someJaxrsMethod(String id);

@PermissionChecker("id")
boolean checkId(long id) {}
```

?

michalvavrik commented on extensions/security/runtime/src/main/java/io/quarkus/security/runtime/QuarkusPermissionSecurityIdentityAugmentor.java:
```diff
@@ -0,0 +1,64 @@
+                    public Uni<Boolean> apply(Permission requiredpermission) {
+                        if (requiredpermission instanceof QuarkusPermission<?> quarkusPermission) {
+                            return quarkusPermission
+                                    .isGranted(identity, blockingExecutor)
```
> I don't mind it given that it is an internal detail right now, but IMHO we should not proceed later with offering this option as an easier to understand augmented permission checker.

Sure, I think you were right there as checkers are fine to use. Maybe we can allow static checker methods on plain java classes, but only if there is ever demand from users.

michalvavrik replied:
Also @sberyozkin , when I suggested it, I didn't know where I will iterate to, this looks differently than my original plan (that was simpler).

sberyozkin replied:
@michalvavrik What I meant was that right now this solution is about the fact that augmenting the identity with permission checking functions is the only way to enforce permission based control, and here we do what looks nicer to the eye, but also block `implies`, the solution is constrained by the fact that the only way to do it is to use permission checker functions.
IMHO, adding the permission checkers should be avoidable even if no `@PermissionChecker` is used, as proposed in #43717. If #43717 is done next, then `QuarkusPermission`, `QuarkusPermissionSecurittyIdentityAugmentor` will probably become redundant, Quarkus security can just do 

```
for (Permission possessedPermission : identity.permissions()) {
possessedPermission.implies(requiredPermission);
}
```
somewhere next to the role based check.
Users will only see in the docs that they need to augment identity with possessed permissions, and not write correct permission checks, this is not what they should do...
Which is why IMHO we should keep `QuarkusPermission`, `QuarkusPermissionSecurittyIdentityAugmentor` low profile for now

michalvavrik replied:
I agree with that and I did act upon that. Only `QuarkusPermission` is public but that cannot be avoided because I wanted to support declaring of permission checker methods on package-private CDI method beans. I think current state is ok and we should not proceed with my original proposal regarding `QuarkusPermission`.

sberyozkin reviewed (approved):
This is going to be a massive feature, thanks @michalvavrik

sberyozkin commented:
@michalvavrik Can I suggest to update the subject a bit to highlight it, something like
`Add PermissionChecker annotation to allow using custom CDI bean methods as permission checkers` ?

michalvavrik commented:
> @michalvavrik Can I suggest to update the subject a bit to highlight it, something like `Add PermissionChecker annotation to allow using custom CDI bean methods as permission checkers` ?

Changed PR title and fixed the QuarkusPermission javadoc.

FroMage commented on docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc:
```diff
@@ -1055,8 +1055,79 @@ Because `MediaLibrary` is the `TvLibrary` class parent, a user with the `admin`
+    }
+}
+----
+<1> Permission required to access the `ProjectResource#renameProject` is the `rename-project` permission.
```
```suggestion
<1> The permission required to access the `ProjectResource#renameProject` is the `rename-project` permission.
```

FroMage commented on docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc:
```diff
@@ -1055,8 +1055,79 @@ Because `MediaLibrary` is the `TvLibrary` class parent, a user with the `admin`
+}
+----
+<1> Permission required to access the `ProjectResource#renameProject` is the `rename-project` permission.
+<2> The `ProjectResource#canRenameProject` method authorize access to the `ProjectResource#renameProject` endpoint.
```
```suggestion
<2> The `ProjectResource#canRenameProject` method authorizes access to the `ProjectResource#renameProject` endpoint.
```

FroMage commented on docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc:
```diff
@@ -1155,47 +1225,31 @@ public class SimpleBeanParam {
 That is particularly useful with Java records, where generated accessors are not prefixed with `get`.
 
-Here is an example of the `BeanParamPermission` permission that checks user principal, custom header and query parameter:
+Here is an example of a `@PermissionChecker` method that checks `say-hello` permission based on a user principal, custom header and query parameter:
```
```suggestion
Here is an example of a `@PermissionChecker` method that checks the `say-hello` permission based on a user principal, custom header and query parameter:
```

FroMage reviewed (approved):
I didn't check the code, just the docs, but the feature is great :)

sberyozkin commented:
Hi @cescoffier, @gsmet, please review this PR, here there will be a few more validators put in place to avoid some possible user errors. 
https://github.com/quarkusio/quarkus-security/pull/56 is only adding an interface

cescoffier commented on bom/application/pom.xml:
```diff
@@ -182,7 +182,7 @@
         <mockito.version>5.14.1</mockito.version>
         <jna.version>5.8.0</jna.version><!-- should satisfy both testcontainers and mongodb -->
-        <quarkus-security.version>2.1.0</quarkus-security.version>
+        <quarkus-security.version>2.1.1-SNAPSHOT</quarkus-security.version>
```
This would need to be fixed.

cescoffier commented on docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc:
```diff
@@ -1055,8 +1055,79 @@ Because `MediaLibrary` is the `TvLibrary` class parent, a user with the `admin`
+        project.name = newName;
+    }
+
+    @PermissionChecker("rename-project") <2>
```
Does it has to be in the same bean? I would prefer an example with 2 different beans.

What about scope BTW (when it's 2 different beans)?

michalvavrik replied:
> Does it has to be in the same bean? I would prefer an example with 2 different beans.

It doesn't need to be same bean. Couple of lines below is an example that shows that.  I can drop this one because it is not important for me at all. When I write docs, I try to show all that is possible and this just looked nice in my eyes. If you require dropping this, I will do that.

> What about scope BTW (when it's 2 different beans)?

Before security checks are started (that is not specific for this method), CDI request scope is activated. It is same request scope.

cescoffier replied:
Ok, I think it should explain that it could be in a different bean (and point to the next example, which I've seen 5 min later). 

About the scope, it should also be mentioned that you can rely on the request scope. 

BTW, is the bean providing the checker necessarily application scoped or singleton? What about dependent, or request scoped?

michalvavrik replied:
> Ok, I think it should explain that it could be in a different bean (and point to the next example, which I've seen 5 min later).
> About the scope, it should also be mentioned that you can rely on the request scope.

It makes sense, I'll do it.

> BTW, is the bean providing the checker necessarily application scoped or singleton? What about dependent, or request scoped?

There is validation in place, it must be either `Singleton`, `ApplicationScoped` or `RequestScoped`. I decided against `Dependent` because `ArcContainer` Javadoc says  that you must explicitly destroy dependent beans when they are accessed programmatically and there reliable hook when `permission` is not going to be used anymore. I think we could add it later if there is sound reason.

michalvavrik replied:
I'll mention that in docs as well.

cescoffier commented on docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc:
```diff
@@ -1055,8 +1055,79 @@ Because `MediaLibrary` is the `TvLibrary` class parent, a user with the `admin`
+[[permission-checker]]
+==== Create permission checkers
+
+By default, `SecurityIdentity` must be augmented with permissions which can be used to check if this identity passes `@PermissionAllowed` authorization restrictions.
```
I don't think using "augmented" is great here. (Augmented is the augmentation phase of Quarkus, and it's in the log of each build). "configured" might be better.

michalvavrik replied:
it can _only_ (technically that's not true, but users don't use http perms for that I think) happen in `SecurityIdentityAugmentor`, that is why I used `augmented`. I am perfectly fine with `configured`, I'll use it.

cescoffier replied:
Thanks! I think _configure_ is less confusing in this context.

I now remember about the `SecurityIdentityAugmentor` discussions.... At that time, everything looked like an augmentation :-)

cescoffier commented on docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc:
```diff
@@ -1055,8 +1055,79 @@ Because `MediaLibrary` is the `TvLibrary` class parent, a user with the `admin`
+==== Create permission checkers
+
+By default, `SecurityIdentity` must be augmented with permissions which can be used to check if this identity passes `@PermissionAllowed` authorization restrictions.
+Alternatively, you can use a `@PermissionChecker` annotation to mark any CDI bean method as a permission checker.
```
Somewhere you need to say that the name must match.

cescoffier commented on docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc:
```diff
@@ -1055,8 +1055,79 @@ Because `MediaLibrary` is the `TvLibrary` class parent, a user with the `admin`
+}
+----
+<1> A CDI bean with the permission checker must be either a normal scoped bean or a `@Singleton` bean.
+<2> The permission checker method must return either `boolean` or `Uni<Boolean>`. Private checker methods are not supported.
```
How are failures handled?

michalvavrik replied:
Failures result in forbidden exception [because only authenticated identities are authorized by these methods). Just FYI, there is a test for that. Exceptions thrown by the check method are logged, but are wrapped.

cescoffier commented on docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc:
```diff
@@ -1055,8 +1055,79 @@ Because `MediaLibrary` is the `TvLibrary` class parent, a user with the `admin`
+<1> A CDI bean with the permission checker must be either a normal scoped bean or a `@Singleton` bean.
+<2> The permission checker method must return either `boolean` or `Uni<Boolean>`. Private checker methods are not supported.
+
+TIP: Permission checks are run on event loops whenever possible.
```
By default, instead of "whenever possible", except if you implement some smart thread introspection.

cescoffier reviewed (changes requested):
I saw in a comment that this feature could replace a Spring Security feature - but I don't see any example / comparison.

cescoffier commented:
> So, should we mention Spring Security inside our guide? I'd say no, but that is just my personal opinion, if you feel strongly about that I don't mind adding this comparison / example @cescoffier .

I agree it should not be in that guide, I was just looking to compare. A blog post would be nice!

michalvavrik commented:
I've addressed @FroMage and @cescoffier comments (thanks!) and added required validations and their test. Specifically I believe `extensions/resteasy-reactive/rest/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/security/PermissionCheckerOnEndpointValidationFailureTest.java` addresses @gsmet concern that `@PermissionChecker` and `@PermissionsAllowed` could be confused with each other. I also pinged Guillaume and asked for Quarkus Security release when the time is right for him.

michalvavrik commented:
Although I made changes when there was electricity outage, now that I rerun all the test, there was ugly hack on `ExecutionModelAnnotationsProcessor` that I had to workaround so that this validation would work. It works now, I am not happy about this though.

michalvavrik commented:
No sorry, `ExecutionModelAnnotationsAllowedBuildItem` is not reliable check whether method is an entrypoint or not as it has false positives for checkers declared on resources. I'll just drop that validation.

michalvavrik committed:
Support @PermissionChecker CDI bean methods

michalvavrik commented:
CI is green!

sberyozkin commented:
Thanks @michalvavrik for completing this important enhancement.

@gsmet Are you still somewhat concerned with how it will all stand against the code evolution or have some other concerns ? I think we are ready to merge otherwise as @cescoffier has also approved. 

Note we hope some internal code made to support this PR may go, with more simplifications planned

michalvavrik commented:
> Thanks @michalvavrik for completing this important enhancement.
> 
> @gsmet Are you still somewhat concerned with how it will all stand against the code evolution or have some other concerns ? I think we are ready to merge otherwise as @cescoffier has also approved.
> 
> Note we hope some internal code made to support this PR may go, with more simplifications planned

I think @gsmet won't be around for next 4 days starting tomorrow. It has been 2 weeks since we discussed it here https://github.com/quarkusio/quarkus-security/pull/56.

While I am personally fine with waiting for whatever time, I don't believe it is necessary. It takes but a minute to give heads up that you need more time to review. I will try to address any concerns raised whether it is now or when PR is merged. Thanks

michalvavrik commented:
That said, let's wait, I didn't mean to press you :-)

sberyozkin commented:
Sure, let me merge, @gsmet, please ping us any time to clarify the details, a brief summary of the PR:
* Any `@PermissionAllowed` restriction on a JAX-RS endpoint can be approved/denied with a boolean method which accepts one or more request parameters and/or the already verified security identity and is annotated with a matching `@PermissionChecker`.
* @michalvavrik added a lot of tests. 2 specific tests which may help address your earlier concerns are about failing when 1) `@PermissionChecker` is attached by mistake to the JAX-RS method and 2) `@PermissionChecker` does not have a matching `@PermissionAllowed` (correctly placed on the JAX-RS method)
